app/tests/heartbeat_controller_desktop_test.cc

```diff
@@ -0,0 +1,222 @@
+  std::string day2 = "2000-01-24";
+  std::string day3 = "2000-01-25";
+  EXPECT_CALL(mock_date_provider_, GetDate()).Times(3)
+    .WillOnce(Return(day1)).WillOnce(Return(day2)).WillOnce(Return(day3));
+
+  // Log a heartbeat for UserAgent1 on day1
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent1);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // Log a heartbeat for UserAgent2 on day2
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent2);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // Log a heartbeat for UserAgent1 on day3
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent1);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // Read from the storage class to verify
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage_.ReadTo(read_heartbeats);
+  ASSERT_TRUE(read_ok);
+  EXPECT_EQ(read_heartbeats.last_logged_date, day3);
+  ASSERT_EQ(read_heartbeats.heartbeats.size(), 2);
+  // UserAgent1 was logged on day1 and day3
+  ASSERT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1].size(), 2);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1][0], day1);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1][1], day3);
+  // UserAgent2 was logged on day2
+  ASSERT_EQ(read_heartbeats.heartbeats[kCustomUserAgent2].size(), 1);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent2][0], day2);
+}
+
+TEST_F(HeartbeatControllerDesktopTest, LogMoreThan30DaysRemovesOldEntries) {
+  {
+    // InSequence guarantees that all of the expected calls occur in order.
+    testing::InSequence seq;
+    for (int month = 1; month <= 3; month++) {
+      for (int day = 1; day <= 30; day++) {
+        // YYYY-MM-DD\0
+        char date[11];
+        snprintf(date, sizeof(date), "2000-%02d-%02d", month, day);
+        std::string date_string(date);
+        EXPECT_CALL(mock_date_provider_, GetDate()).WillOnce(Return(date_string));
+      }
+    }
+  }
+  for (int i = 1; i <= 90; i++) {
+    controller_.LogHeartbeat();
+  }
+  // Since LogHeartbeat is done asynchronously, wait a bit before verifying that the log succeeded.
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // Read from the storage class to verify
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage_.ReadTo(read_heartbeats);
+  ASSERT_TRUE(read_ok);
+  EXPECT_EQ(read_heartbeats.last_logged_date, "2000-03-30");
+  ASSERT_EQ(read_heartbeats.heartbeats.size(), 1);
+  // Even though heartbeat logging is asynchronous, it happens in the order that it is scheduled
+  // So the heartbeats should be logged in the correct order.
+  for (auto const& entry : read_heartbeats.heartbeats) {
+    std::vector<std::string> dates = entry.second;
+    ASSERT_EQ(dates.size(), 30);
+    EXPECT_EQ(dates[0], "2000-03-01");
+    EXPECT_EQ(dates[29], "2000-03-30");
+  }
+}
+
```

Comment:
Is there plan to add more edge case here?
For instance:
- What if log `2000-01-25` then `2000-01-24`?
- Create destroy tests for race condition
- Having two LoggedHeartbeats instances pointing to the same app_id
etc.

Reply:
Added a test for logging dates in reverse-chronological order and verifying that only the newest date is logged.

Added a test that calls LogHeartbeat many times and then calls the controller's Destructor
- note: current implementation of Scheduler's destructor is to stop allowing scheduling of new work and to join with the worker thread. So all already-scheduled work is actually completed.

Added a test with multiple controllers for the same app id trying to log heartbeats simultaneously. Confirmed that it currently fails due to concurrency issues of read/write happening between other read/write operations, so the test is disabled.
Filed b/237003018 to track this